### PR TITLE
Salesforce reduce api calls

### PR DESCRIPTION
**Pull request by isleshocky77**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | N
| Related user documentation PR URL | N/A
| Related developer documentation PR URL | N/A
| Issues addressed (#s or URLs) | #2836 
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This pull request reduces in greatly the number of Salesfore API calls used for a fetch leads call. In my testing of it when importing a duration of 29 days which had 18 leads it reduced the number of API requests from 27 to 5 on the first callling of fetch leads. On all subsequent calls it went from 27 to 2 API requests.

#### Steps to test this PR:
1. run `./app/console -vvv mautic:integration:fetchleads -i "Salesforce" -a "29 day"`
2. use a debugger to watch /plugins/MauticCrmBundle/Api/SalesforceApi.php#L53 ($this->integration->makeRequest())
3. Use code from this request and repeat steps 1 and 2

#### Notes:

In my opinion I don't love the solution I've given here. I would prefer that we cache the organization createdDate when authorizing the credentials. I would also prefer re-writing the `SalesforceApi::getLeads()` to build the query from the "matched" fields which are made during setting up the integration. You could catch errors with the query if that ever happens.

With that said, this is what I currently had time for and made a massive improvement in the number of API calls. I'm putting the pull request out there in the case anyone else would like to use this code and/or improve on it.

### CODE REVIEW

File: plugins/MauticCrmBundle/Integration/SalesforceIntegration.php [hunk 1]
```diff
@@ -827,4 +838,16 @@ public function getLeadData(\DateTime $startDate = null, \DateTime $endDate = nu
...
     }
+
+    /**
+     * @return FilesystemAdapter
+     *
+     * @TODO This should probably be moved to a service and system-wide
+     */
+    public function getCache()
+    {
+        $cacheDir = $this->dispatcher->getContainer()->getParameter('kernel.cache_dir');
```
Comment (mqueme) [inline plugins/MauticCrmBundle/Integration/SalesforceIntegration.php]: Please change this line to:

`$this->dispatcher->getContainer()->get('mautic.helper.paths')->getSystemPath('cache')`

**Comment by mqueme** [comment]
Is there a way to use database cache instead of a filesystem cache, with this?

**Comment by isleshocky77** [comment]
@mqueme You could use any caching. If you have an example of a current pattern for caching that is being used throughout Mautic I would be fine switching to that. For now this was quick and clean way for me to cache the API calls which I knew I would not need to keep making. As I said in the above comments. I would actually recommend re-writing a bunch of the workflow of the integration to make a bit more sense, but this is a quick easy solution which cut down the calls from 27 to 2 without losing any functionality.

**Comment by isleshocky77** [comment]
@alanhartless I've rebased this on staging (2.5.2-dev). I also removed some of the commits which weren't exactly related to this PR (warning fixes).

**Comment by isleshocky77** [comment]
@alanhartless What type of documentation are you looking for and which repository? There is nothing new to this PR except enhancing the efficiency of the already existing feature. Perhaps you may no better how this should be documented for Mautic purposes for the caching changes you made.

**Comment by alanhartless** [comment]
Sorry - that was more of a label for myself to document the caching side of things in the developer docs.

**Comment by isleshocky77** [comment]
Rebased on staging after #3167 merge for CI testing

**Comment by mqueme** [comment]
I think this is unrelated to this PR but, I am getting some errors regarding the owner, when fetching companies, and also some column errors when fetching contacts. Besides that, cache works.

**Comment by isleshocky77** [comment]
@mqueme can you put the owner errors in another issue so I can take a look and see if I'm having the same issue and correct it before next release?

**Comment by alanhartless** [comment]
@mqueme @isleshocky77 it was related to the cached fields. Should be fixed now.

**Comment by mqueme** [comment]
With new changes, this now runs without errors, cache is being saved on the DB

**Pull request merged**